### PR TITLE
HDDS-3882. Update modification time when updating volume/bucket/key ACLs

### PR DESCRIPTION
**Pull request by cxorm**
## What changes were proposed in this pull request?
For renewing modification time when updating Acls of `Volume`/ `Bucket`/ `Key`, there are three parts to be modified in this patch.

**I. Volume Part**
  - addacl: Add one or more new ACLs.
  - removeacl: Remove one or more existing ACLs.
  - setacl: Set one or more ACLs, replacing the existing ones.
    
**II. Bucket Part**
  - addacl
  - removeacl
  - setacl

**III. Key Part**
  - addacl
  - removeacl
  - setacl

#### Proposed fix.
Let `preExecute()` be overrided by `OMAddAclRequest`/ `OMSetAclRequest`/ `OMRemoveAclRequest` of `Volume`/ `Bucket`/`Key`, and evalute modification time in `preExecute()`.

After `preExecute()`, renew modification time after true `acl apply operation` in `validateAndUpdateCache()` of `OMVolumeAclRequest`/ `OMBucketAclRequest`/ `OMKeyAclRequest`.

**[note]**
In `OMVolumeAclRequest` and `OMBucketAclRequest`, the cache of table is updated only when true `acl apply operation` happening, so the code snippet of renewing modification time is added in this block.

But in `OMKeyAclRequest`, the original flow didn't consider boolean of `acl apply operation` when updating the cache of table, so the if-condition includes boolean of `acl apply operation`.

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-3882

## How was this patch tested?
- Modify/Add unit tests.
  Volume Part : Just modify the `testPreExecute()` to verify modification time is renewed.
      
  Bucket Part : Add new unit tests to verify `OMBucketAddAclRequest`/ `OMBucketRemoveAclRequest`/ `OMBucketSetAclRequest` including `testPreExecute()` and `testValidateAndUpdateCache`.
  (Feel free to correct me if anyone think this part should be addressed in separated Jira or any thoughts :) )
  
  Key Part : Modify original `TestOMKeyAclRequest.java` to `testKeyRemoveAclRequest()` and `testKeySetAclRequest()`.
  (Again, Feel free to correct me if anyone think this part should be addressed in separated Jira or any thoughts, thanks)
 

- [Github action passed except codecov](https://github.com/cxorm/hadoop-ozone/runs/957353259)

### CODE REVIEW

**Comment by cxorm** [comment]
Test would be added to pass code coverage.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/bucket/acl/OMBucketAddAclRequest.java [hunk 1]
```diff
@@ -58,6 +60,19 @@
     };
   }
 
+  @Override
+  public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
+    long modificationTime = Time.now();
+    OzoneManagerProtocolProtos.AddAclRequest addAclRequest =
+        getOmRequest().getAddAclRequest().toBuilder()
+            .setModificationTime(modificationTime).build();
```
Comment (xiaoyuyao) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/bucket/acl/OMBucketAddAclRequest.java:68]: if you don't call .build() hereon line 68, we can avoid the unnecessary  toBuilder() conversion on line 71.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/bucket/acl/OMBucketRemoveAclRequest.java [hunk 2]
```diff
@@ -55,6 +57,19 @@
     };
   }
 
+  @Override
+  public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
+    long modificationTime = Time.now();
+    OzoneManagerProtocolProtos.RemoveAclRequest removeAclRequest =
+        getOmRequest().getRemoveAclRequest().toBuilder()
```
Comment (xiaoyuyao) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/bucket/acl/OMBucketRemoveAclRequest.java:64]: same as above.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/bucket/acl/OMBucketSetAclRequest.java [hunk 3]
```diff
@@ -55,6 +57,19 @@
     };
   }
 
+  @Override
+  public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
+    long modificationTime = Time.now();
+    OzoneManagerProtocolProtos.SetAclRequest setAclRequest =
+        getOmRequest().getSetAclRequest().toBuilder()
```
Comment (xiaoyuyao) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/bucket/acl/OMBucketSetAclRequest.java:64]: same as above.

**Review by xiaoyuyao — COMMENTED** [review]
Thanks @cxorm for the update. The latest PR LGTM overall, a few comments added inline.

**Comment by cxorm** [comment]
> Thanks @cxorm for the update. The latest PR LGTM overall, a few comments added inline.

Thank you @xiaoyuyao for the careful review.
The fix was added to addressed the comments.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeyAddAclRequest.java [hunk 4]
```diff
@@ -43,6 +45,19 @@
   private static final Logger LOG =
       LoggerFactory.getLogger(OMKeyAddAclRequest.class);
 
+  @Override
+  public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
+    long modificationTime = Time.now();
+    OzoneManagerProtocolProtos.AddAclRequest addAclRequest =
+        getOmRequest().getAddAclRequest().toBuilder()
+        .setModificationTime(modificationTime).build();
```
Comment (xiaoyuyao) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeyAddAclRequest.java:53]: This has the same toBuilder=>build=>toBuilder issue.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeyRemoveAclRequest.java [hunk 5]
```diff
@@ -43,6 +45,19 @@
   private static final Logger LOG =
       LoggerFactory.getLogger(OMKeyRemoveAclRequest.class);
 
+  @Override
+  public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
+    long modificationTime = Time.now();
+    OzoneManagerProtocolProtos.RemoveAclRequest removeAclRequest =
+        getOmRequest().getRemoveAclRequest().toBuilder()
+            .setModificationTime(modificationTime).build();
```
Comment (xiaoyuyao) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeyRemoveAclRequest.java:53]: Same as above.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeySetAclRequest.java [hunk 6]
```diff
@@ -44,6 +46,19 @@
   private static final Logger LOG =
       LoggerFactory.getLogger(OMKeySetAclRequest.class);
 
+  @Override
+  public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
+    long modificationTime = Time.now();
+    OzoneManagerProtocolProtos.SetAclRequest setAclRequest =
+        getOmRequest().getSetAclRequest().toBuilder()
```
Comment (xiaoyuyao) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/OMKeySetAclRequest.java:53]: Same as above. Sorry, I did not explicitly comment all on this.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/volume/acl/OMVolumeAddAclRequest.java [hunk 7]
```diff
@@ -48,6 +50,19 @@
     volumeAddAclOp = (acls, volArgs) -> volArgs.addAcl(acls.get(0));
   }
 
+  @Override
+  public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
+    long modificationTime = Time.now();
+    OzoneManagerProtocolProtos.AddAclRequest addAclRequest =
+        getOmRequest().getAddAclRequest().toBuilder()
```
Comment (xiaoyuyao) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/volume/acl/OMVolumeAddAclRequest.java:57]: Same as above.

**Review by xiaoyuyao — COMMENTED** [review]
Thanks for the update @cxorm. A few similar usage that can be optimized. Sorry I did not explicitly comment all of them last time.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/volume/acl/OMVolumeAclRequest.java [hunk 8]
```diff
@@ -84,6 +84,7 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
           VOLUME_LOCK, volume);
       omVolumeArgs = getVolumeInfo(omMetadataManager, volume);
 
+
```
Comment (cxorm) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/volume/acl/OMVolumeAclRequest.java:87]: nit

**Comment by cxorm** [comment]
> Thanks for the update @cxorm. A few similar usage that can be optimized. Sorry I did not explicitly comment all of them last time.

That's my carelessness : )
Thanks a lot @xiaoyuyao for reviewing it again, the toBuilder issues in AclRequest has been addressed.

**Comment by xiaoyuyao** [comment]
Thanks @cxorm  for fixing this. The change LGTM, +1. I will merge it shortly. 

Also notice similar unnecessary builder conversion issue in OMVolumeSetQuotaRequestwhen the modify time is updated. Can you check and open separate JIRA to fix those? Thanks!

**Comment by cxorm** [comment]
> Thanks @cxorm for fixing this. The change LGTM, +1. I will merge it shortly.
> 
> Also notice similar unnecessary builder conversion issue in OMVolumeSetQuotaRequestwhen the modify time is updated. Can you check and open separate JIRA to fix those? Thanks!

Thank you @xiaoyuyao for reviewing it again and again and committing it.
I would check the issues mentioned.

**Pull request merged**